plugins/cfu/fu-cfu-module.c

```diff
@@ -0,0 +1,110 @@
+/*#
+ * Copyright (C) 2021 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <fwupdplugin.h>
+
+#include "fu-cfu-module.h"
+
+struct _FuCfuModule {
+	FuDevice parent_instance;
+	guint8 component_id;
+	guint8 bank;
+};
+
+G_DEFINE_TYPE(FuCfuModule, fu_cfu_module, FU_TYPE_DEVICE)
+
+static void
+fu_cfu_module_to_string(FuDevice *device, guint idt, GString *str)
+{
+	FuCfuModule *self = FU_CFU_MODULE(device);
+	fu_common_string_append_kx(str, idt, "Bank", self->bank);
+	fu_common_string_append_kx(str, idt, "ComponentId", self->component_id);
+}
+
+gboolean
+fu_cfu_module_setup(FuCfuModule *self, const guint8 *buf, gsize bufsz, gsize offset, GError **error)
+{
+	FuDevice *parent = fu_device_get_proxy(FU_DEVICE(self));
+	guint32 version_raw = 0;
+	guint8 tmp = 0;
+	g_autofree gchar *instance_id0 = NULL;
+	g_autofree gchar *instance_id1 = NULL;
+	g_autofree gchar *instance_id2 = NULL;
+	g_autofree gchar *logical_id = NULL;
+	g_autofree gchar *name = NULL;
+	g_autofree gchar *version = NULL;
+
+	/* component ID */
+	if (!fu_common_read_uint8_safe(buf, bufsz, offset + 0x5, &self->component_id, error))
+		return FALSE;
+	name = g_strdup_printf("%s (0x%02X)", fu_device_get_name(parent), self->component_id);
+	fu_device_set_name(FU_DEVICE(self), name);
+
+	/* these GUIDs may cause the name or version-format to be overwritten */
+	instance_id0 = g_strdup_printf("USB\\VID_%04X&PID_%04X",
+				       fu_udev_device_get_vendor(FU_UDEV_DEVICE(parent)),
+				       fu_udev_device_get_model(FU_UDEV_DEVICE(parent)));
+	fu_device_add_instance_id(FU_DEVICE(self), instance_id0);
+	instance_id1 = g_strdup_printf("USB\\VID_%04X&PID_%04X&CID_%02X",
+				       fu_udev_device_get_vendor(FU_UDEV_DEVICE(parent)),
+				       fu_udev_device_get_model(FU_UDEV_DEVICE(parent)),
+				       self->component_id);
+	fu_device_add_instance_id(FU_DEVICE(self), instance_id1);
+
+	/* bank */
+	if (!fu_common_read_uint8_safe(buf, bufsz, offset + 0x4, &tmp, error))
+		return FALSE;
+	self->bank = tmp & 0b11;
+	instance_id2 = g_strdup_printf("USB\\VID_%04X&PID_%04X&CID_%02X&BANK_%01X",
+				       fu_udev_device_get_vendor(FU_UDEV_DEVICE(parent)),
+				       fu_udev_device_get_model(FU_UDEV_DEVICE(parent)),
+				       self->component_id,
+				       self->bank);
+	fu_device_add_instance_id(FU_DEVICE(self), instance_id2);
```

Comment:
This is a bit odd to me - why would each bank have it's own instance ID?
Also if more than one bank is available the device, should apply `FWUPD_DEVICE_FLAG_DUAL_IMAGE`

Reply:
> why would each bank have it's own instance ID

Assuming there's more than one bag (and that is making an assumption...) The firmware _is_ specific to the bank. e.g. like the wacom-usb payload where there are different entry addresses for each bank firmware. In some cases there's only one bank (of the bank firmware is the same somehow) so I went with adding both to be most flexible.

I think `FWUPD_DEVICE_FLAG_DUAL_IMAGE` might have to be a quirk; the only way we could award that is if we "see" both banks in the GET_VERSION command. I'm guessing we don't actually know if there is:

 * a device with only one bank and the new firmware is loaded into RAM for the CFU process
 * a device with two banks and only the running one is reported, and the new firmware is supposed to be the *other* one.